BUG: `np.unique_inverse` output has wrong shape

### Describe the issue:

According to the [`unique_inverse` specification](https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.unique_inverse.html), the `inverse_indices` output of `np.unique_inverse` should have the same shape as the input:

> second element must have the field name `inverse_indices` and must be an array containing the indices of values that reconstruct `x`. The array must have the same shape as `x` and have the default array index data type.

In NumPy's current implementation, `inverse_indices` is a flattened array.

A similar issue exists for `numpy.unique_all`.

### Reproduce the code example:

```python
import numpy as np
x = np.ones((2, 3))
out = np.unique_inverse(x)
assert out.inverse_indices.shape == x.shape, f"{out.inverse_indices.shape=} {x.shape=}"
```


### Error message:

```shell
AssertionError: out.inverse_indices.shape=(6,) x.shape=(2, 3)
```


### Python and NumPy Versions:
```
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
2.0.0.dev0+git20240106.42fef42
3.9.6 (default, May  7 2023, 23:32:44) 
[Clang 14.0.3 (clang-1403.0.22.14.1)]
```
### Runtime Environment:
```
>>> import numpy; print(numpy.show_runtime())
[{'numpy_version': '2.0.0.dev0+git20240106.42fef42',
  'python': '3.9.6 (default, May  7 2023, 23:32:44) \n'
            '[Clang 14.0.3 (clang-1403.0.22.14.1)]',
  'uname': uname_result(system='Darwin', node='jmdg-macbookpro.roam.internal', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
```

--- a/doc/source/release/2.0.0-notes.rst
+++ b/doc/source/release/2.0.0-notes.rst
@@ -313,3 +313,12 @@ precalculated mean as an keyword argument. See the doc-strings for details and a
 example illustrating the speed-up.
 
 (`gh-24126 <https://github.com/numpy/numpy/pull/24126>`__)
+
+``np.unique`` ``return_inverse`` shape for multi-dimensional inputs
+-------------------------------------------------------------------
+When multi-dimensional inputs are passed to ``np.unique`` with ``return_inverse=True``,
+the ``unique_inverse`` output is now shaped such that the input can be reconstructed
+directly using ``np.take(unique, unique_inverse)`` when ``axis = None``, and
+``np.take_along_axis(unique, unique_inverse, axis=axis)`` otherwise.
+
+(`gh-25553 <https://github.com/numpy/numpy/pull/24126>`__)

--- a/numpy/lib/_arraysetops_impl.py
+++ b/numpy/lib/_arraysetops_impl.py
@@ -212,7 +212,7 @@ def unique(ar, return_index=False, return_inverse=False,
     flattened subarrays are sorted in lexicographic order starting with the
     first element.
 
-    .. versionchanged: NumPy 1.21
+    .. versionchanged: 1.21
         If nan values are in the input array, a single nan is put
         to the end of the sorted unique values.
 
@@ -222,7 +222,7 @@ def unique(ar, return_index=False, return_inverse=False,
         lexicographical order is chosen - see np.sort for how the lexicographical
         order is defined for complex arrays.
 
-    .. versionchanged: NumPy 2.0
+    .. versionchanged: 2.0
         For multi-dimensional inputs, ``unique_inverse`` is reshaped such that
         the input can be reconstructed using ``np.take(unique, unique_inverse)``
         when ``axis = None``, and

--- a/numpy/lib/_arraysetops_impl.py
+++ b/numpy/lib/_arraysetops_impl.py
@@ -223,9 +223,9 @@ def unique(ar, return_index=False, return_inverse=False,
         order is defined for complex arrays.
 
     .. versionchanged: 2.0
-        For multi-dimensional inputs, ``unique_inverse`` is reshaped such that
-        the input can be reconstructed using ``np.take(unique, unique_inverse)``
-        when ``axis = None``, and
+        For multi-dimensional inputs, ``unique_inverse`` is reshaped
+        such that the input can be reconstructed using
+        ``np.take(unique, unique_inverse)`` when ``axis = None``, and
         ``np.take_along_axis(unique, unique_inverse, axis=axis)`` otherwise.
 
     Examples

--- a/numpy/lib/_arraysetops_impl.py
+++ b/numpy/lib/_arraysetops_impl.py
@@ -422,14 +422,17 @@ def unique_all(x):
     unique : Find the unique elements of an array.
 
     """
-    result = unique(
+    x = np.asanyarray(x)
+    values, indices, inverse_indices, counts = unique(
         x,
         return_index=True,
         return_inverse=True,
         return_counts=True,
         equal_nan=False
     )
-    return UniqueAllResult(*result)
+    inverse_indices = inverse_indices.reshape(x.shape)
+    return UniqueAllResult(values=values, indices=indices,
+                           inverse_indices=inverse_indices, counts=counts)
 
 
 def _unique_counts_dispatcher(x, /):
@@ -509,14 +512,16 @@ def unique_inverse(x):
     unique : Find the unique elements of an array.
 
     """
-    result = unique(
+    x = np.asanyarray(x)
+    values, inverse_indices = unique(
         x,
         return_index=False,
         return_inverse=True,
         return_counts=False,
         equal_nan=False
     )
-    return UniqueInverseResult(*result)
+    inverse_indices = inverse_indices.reshape(x.shape)
+    return UniqueInverseResult(values=values, inverse_indices=inverse_indices)
 
 
 def _unique_values_dispatcher(x, /):

--- a/numpy/lib/_arraysetops_impl.py
+++ b/numpy/lib/_arraysetops_impl.py
@@ -222,6 +222,12 @@ def unique(ar, return_index=False, return_inverse=False,
         lexicographical order is chosen - see np.sort for how the lexicographical
         order is defined for complex arrays.
 
+    .. versionchanged: NumPy 2.0
+        For multi-dimensional inputs, ``unique_inverse`` is reshaped such that
+        the input can be reconstructed using ``np.take(unique, unique_inverse)``
+        when ``axis = None``, and
+        ``np.take_along_axis(unique, unique_inverse, axis=axis)`` otherwise.
+
     Examples
     --------
     >>> np.unique([1, 1, 2, 2, 3, 3])
@@ -273,7 +279,7 @@ def unique(ar, return_index=False, return_inverse=False,
     ar = np.asanyarray(ar)
     if axis is None:
         ret = _unique1d(ar, return_index, return_inverse, return_counts, 
-                        equal_nan=equal_nan)
+                        equal_nan=equal_nan, inverse_shape=ar.shape)
         return _unpack_tuple(ret)
 
     # axis was specified and not None
@@ -282,6 +288,8 @@ def unique(ar, return_index=False, return_inverse=False,
     except np.exceptions.AxisError:
         # this removes the "axis1" or "axis2" prefix from the error message
         raise np.exceptions.AxisError(axis, ar.ndim) from None
+    inverse_shape = [1] * ar.ndim
+    inverse_shape[axis] = ar.shape[0]
 
     # Must reshape to a contiguous 2D array for this to work...
     orig_shape, orig_dtype = ar.shape, ar.dtype
@@ -316,13 +324,14 @@ def unique(ar, return_index=False, return_inverse=False,
         return uniq
 
     output = _unique1d(consolidated, return_index,
-                       return_inverse, return_counts, equal_nan=equal_nan)
+                       return_inverse, return_counts,
+                       equal_nan=equal_nan, inverse_shape=inverse_shape)
     output = (reshape_uniq(output[0]),) + output[1:]
     return _unpack_tuple(output)
 
 
 def _unique1d(ar, return_index=False, return_inverse=False,
-              return_counts=False, *, equal_nan=True):
+              return_counts=False, *, equal_nan=True, inverse_shape=None):
     """
     Find the unique elements of an array, ignoring shape.
     """
@@ -359,7 +368,7 @@ def _unique1d(ar, return_index=False, return_inverse=False,
         imask = np.cumsum(mask) - 1
         inv_idx = np.empty(mask.shape, dtype=np.intp)
         inv_idx[perm] = imask
-        ret += (inv_idx,)
+        ret += (inv_idx.reshape(inverse_shape),)
     if return_counts:
         idx = np.concatenate(np.nonzero(mask) + ([mask.size],))
         ret += (np.diff(idx),)
@@ -422,17 +431,14 @@ def unique_all(x):
     unique : Find the unique elements of an array.
 
     """
-    x = np.asanyarray(x)
-    values, indices, inverse_indices, counts = unique(
+    result = unique(
         x,
         return_index=True,
         return_inverse=True,
         return_counts=True,
         equal_nan=False
     )
-    inverse_indices = inverse_indices.reshape(x.shape)
-    return UniqueAllResult(values=values, indices=indices,
-                           inverse_indices=inverse_indices, counts=counts)
+    return UniqueAllResult(*result)
 
 
 def _unique_counts_dispatcher(x, /):
@@ -512,16 +518,14 @@ def unique_inverse(x):
     unique : Find the unique elements of an array.
 
     """
-    x = np.asanyarray(x)
-    values, inverse_indices = unique(
+    result = unique(
         x,
         return_index=False,
         return_inverse=True,
         return_counts=False,
         equal_nan=False
     )
-    inverse_indices = inverse_indices.reshape(x.shape)
-    return UniqueInverseResult(values=values, inverse_indices=inverse_indices)
+    return UniqueInverseResult(*result)
 
 
 def _unique_values_dispatcher(x, /):

--- a/numpy/lib/tests/test_arraysetops.py
+++ b/numpy/lib/tests/test_arraysetops.py
@@ -960,3 +960,4 @@ class TestUnique:
             result = func(arr)
             assert_array_equal(expected_values, result.values)
             assert_array_equal(expected_inverse, result.inverse_indices)
+            assert_array_equal(arr, result.values[result.inverse_indices])

--- a/numpy/lib/tests/test_arraysetops.py
+++ b/numpy/lib/tests/test_arraysetops.py
@@ -686,7 +686,7 @@ class TestUnique:
         # test for ticket #4785
         a = [(1, 2), (1, 2), (2, 3)]
         unq = [1, 2, 3]
-        inv = [0, 1, 0, 1, 1, 2]
+        inv = [[0, 1], [0, 1], [1, 2]]
         a1 = unique(a)
         assert_array_equal(a1, unq)
         a2, a2_inv = unique(a, return_inverse=True)
@@ -810,6 +810,16 @@ class TestUnique:
         uniq = unique(x, axis=axis)
         assert_array_equal(uniq, [1, 2, 3, 4])
 
+    @pytest.mark.parametrize("axis", [None, 0, -1])
+    def test_unique_inverse_with_axis(self, axis):
+        x = np.array([[4, 4, 3], [2, 2, 1], [2, 2, 1], [4, 4, 3]])
+        uniq, inv = unique(x, return_inverse=True, axis=axis)
+        assert_equal(inv.ndim, x.ndim)
+        if axis is None:
+            assert_array_equal(x, np.take(uniq, inv))
+        else:
+            assert_array_equal(x, np.take_along_axis(uniq, inv, axis=axis))
+
     def test_unique_axis_zeros(self):
         # issue 15559
         single_zero = np.empty(shape=(2, 0), dtype=np.int8)
@@ -820,7 +830,7 @@ class TestUnique:
         assert_equal(uniq.dtype, single_zero.dtype)
         assert_array_equal(uniq, np.empty(shape=(1, 0)))
         assert_array_equal(idx, np.array([0]))
-        assert_array_equal(inv, np.array([0, 0]))
+        assert_array_equal(inv, np.array([[0], [0]]))
         assert_array_equal(cnt, np.array([2]))
 
         # there's 0 elements of shape (2,) along axis 1
@@ -830,7 +840,7 @@ class TestUnique:
         assert_equal(uniq.dtype, single_zero.dtype)
         assert_array_equal(uniq, np.empty(shape=(2, 0)))
         assert_array_equal(idx, np.array([]))
-        assert_array_equal(inv, np.array([]))
+        assert_array_equal(inv, np.empty((1, 0)))
         assert_array_equal(cnt, np.array([]))
 
         # test a "complicated" shape
@@ -899,7 +909,7 @@ class TestUnique:
         msg = "Unique's return_index=True failed with axis=0"
         assert_array_equal(data[idx], uniq, msg)
         msg = "Unique's return_inverse=True failed with axis=0"
-        assert_array_equal(uniq[inv], data)
+        assert_array_equal(np.take_along_axis(uniq, inv, axis=0), data)
         msg = "Unique's return_counts=True failed with axis=0"
         assert_array_equal(cnt, np.array([2, 2]), msg)
 
@@ -908,7 +918,7 @@ class TestUnique:
         msg = "Unique's return_index=True failed with axis=1"
         assert_array_equal(data[:, idx], uniq)
         msg = "Unique's return_inverse=True failed with axis=1"
-        assert_array_equal(uniq[:, inv], data)
+        assert_array_equal(np.take_along_axis(uniq, inv, axis=1), data)
         msg = "Unique's return_counts=True failed with axis=1"
         assert_array_equal(cnt, np.array([2, 1, 1]), msg)
 

--- a/numpy/lib/tests/test_arraysetops.py
+++ b/numpy/lib/tests/test_arraysetops.py
@@ -950,3 +950,13 @@ class TestUnique:
             assert len(res_unique_array_api) == len(res_unique)
             for actual, expected in zip(res_unique_array_api, res_unique):
                 assert_array_equal(actual, expected)
+
+    def test_unique_inverse_shape(self):
+        # Regression test for https://github.com/numpy/numpy/issues/25552
+        arr = np.array([[1, 2, 3], [2, 3, 1]])
+        expected_values, expected_inverse = np.unique(arr, return_inverse=True)
+        expected_inverse = expected_inverse.reshape(arr.shape)
+        for func in np.unique_inverse, np.unique_all:
+            result = func(arr)
+            assert_array_equal(expected_values, result.values)
+            assert_array_equal(expected_inverse, result.inverse_indices)

